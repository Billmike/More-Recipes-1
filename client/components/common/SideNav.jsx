import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import { decodeToken } from '../../utils/authenticate';

const propTypes = {
  handleToggleSignupModal: PropTypes.func,
  handleToggleSigninModal: PropTypes.func,
  handleLogoutUser: PropTypes.func.isRequired,
  user: PropTypes.shape({
    isAuthenticated: PropTypes.bool.isRequired,
    userProfile: PropTypes.shape({
      notifications: PropTypes.bool,
    })
  }).isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string
  }).isRequired,
  userRecipes: PropTypes.shape({
    userAddedRecipesCount: PropTypes.number,
    userFavoritesCount: PropTypes.number,
  }),
  editUserProfile: PropTypes.func.isRequired,
  currentProfileUserId: PropTypes.number,
  authenticatedUserId: PropTypes.number,
};

const defaultProps = {
  handleToggleSignupModal: undefined,
  handleToggleSigninModal: undefined,
  currentProfileUserId: null,
  authenticatedUserId: null,
  userRecipes: undefined,
};

/**
 * Class representing Side nav component
 *
 * @class SideNav
 *
 * @extends {React.Component}
 */
class SideNav extends React.Component {
  /**
   * ComponentDidMount lifecycle method
   *
   * @returns {any} Initialize materialize collapsible class
   *
   * @memberof SideNav
   */
  componentDidMount() {
    $('.collapsible').collapsible();
  }

  /**
   * Handle change in notification switch
   *
   * @returns {object} Change the switch position
   *
   * @memberof SideNav
   */
  handleNotificationChange = () => {
    const { editUserProfile } = this.props;
    const checked = document.getElementById('notification-switch').checked;
    const status = {
      notifications: checked,
    };

    editUserProfile(status);
  }

  /**
   * Render function
   *
   * @returns {object} React element
   *
   * @memberof SideNav
   */
  render() {
    const {
      user, location: { pathname }, userRecipes, handleToggleSignupModal,
      handleToggleSigninModal, handleLogoutUser, currentProfileUserId,
      authenticatedUserId,
    } = this.props;
    const { isAuthenticated, userProfile: { notifications } } = user;
    const userUrl = new RegExp(/users/);
    const dashboard = pathname.match(userUrl) === null ? null : 'fixed';
    const decoded = decodeToken();
    let userId;

    if (decoded) {
      const { user: { id } } = decoded;
      userId = id;
    }
    return (
      <div>
        <ul className={`side-nav ${dashboard}`} id="slide_out">
          <li><Link to="/">Home</Link></li>
          <li>
            <ul className="collapsible" data-collapsible="accordion">
              <li>
                <div className="collapsible-header black-text">Category</div>
                <div className="collapsible-body">
                  <ul>
                    <li>
                      <a href="#!" className="collection-item">Breakfast</a>
                    </li>
                    <li><a href="#!" className="collection-item">Lunch</a></li>
                    <li><a href="#!" className="collection-item">Dinner</a></li>
                    <li>
                      <a href="#!" className="collection-item">Appetizer</a>
                    </li>
                    <li><a href="#!" className="collection-item">Main</a></li>
                    <li>
                      <a href="#!" className="collection-item">Dessert</a>
                    </li>
                  </ul>
                </div>
              </li>
            </ul>
          </li>
          <li>
            <div className="divider" />
          </li>
          {isAuthenticated
            ? <div>
              <li><Link to={`/users/${userId}/dashboard`} >Dashboard</Link></li>
              <li>
                <div className="divider" />
              </li>
              <li>
                <a name="user-recipes" className="subheader">
                  Recipes
                  <span className="badge">
                    {userRecipes && userRecipes.userAddedRecipesCount}
                  </span>
                </a>
              </li>
              <li>
                <a name="user-favorites" className="subheader">
                  Favorites
                  <span className="badge">
                    {userRecipes && userRecipes.userFavoritesCount}
                  </span>
                </a>
              </li>
              {currentProfileUserId === authenticatedUserId &&
                <li>
                  <a>
                    Notificatons
                    <span className=" badge switch">
                      <label htmlFor="notification-switch">
                        <input
                          id="notification-switch"
                          type="checkbox"
                          onClick={this.handleNotificationChange}
                          defaultChecked={notifications}
                        />
                        <span className="lever" />
                      </label>
                    </span>
                  </a>
                </li>
              }
              <li>
                <div className="divider" />
              </li>
              <li>
                <a
                  role="button"
                  tabIndex="0"
                  onClick={handleLogoutUser}
                >
                  Logout
                </a>
              </li>
            </div>
            : <div>
              <li>
                <a
                  role="button"
                  tabIndex="0"
                  onClick={handleToggleSigninModal}
                >
                  Sign In
                </a>
              </li>
              <li>
                <a
                  role="button"
                  tabIndex="0"
                  onClick={handleToggleSignupModal}
                >
                  Register
                </a>
              </li>
            </div>
          }
        </ul>
      </div>
    );
  }
}

SideNav.propTypes = propTypes;
SideNav.defaultProps = defaultProps;

export default SideNav;
