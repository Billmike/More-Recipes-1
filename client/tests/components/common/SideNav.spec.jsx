import React from 'react';

import SideNav from '../../../components/common/SideNav';
import userMockData from '../../__mocks__/user';

describe('SideNav component', () => {
  let spy;

  beforeEach(() => {
    localStorage.setItem(
      'token',
      userMockData.signinSuccessResponse.data.user.token
    );
  });
  afterEach(() => {
    spy.mockClear();
  });
  it('should render component for unauthenticated user', () => {
    const props = {
      handleToggleSignupModal: jest.fn(),
      handleToggleSigninModal: jest.fn(),
      handleLogoutUser: jest.fn(),
      handleSearchCategory: jest.fn(),
      user: {
        isAuthenticated: false,
        userProfile: {
          notifications: false,
        },
      },
      location: {
        pathname: 'path URL',
      },
      userRecipes: {
        userAddedRecipesCount: 0,
        userFavoritesCount: 0,
      },
      editUserProfile: jest.fn(),
      currentProfileUserId: 1,
      authenticatedUserId: 1,
    };
    spy = jest.spyOn(SideNav.prototype, 'componentDidMount');
    const wrapper = shallow(<SideNav {...props} />);

    expect(wrapper).toBeDefined();
    expect(spy).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render component for authenticated user', () => {
    const props = {
      handleToggleSignupModal: jest.fn(),
      handleToggleSigninModal: jest.fn(),
      handleLogoutUser: jest.fn(),
      handleSearchCategory: jest.fn(),
      user: {
        isAuthenticated: true,
        userProfile: {
          notifications: false,
        },
      },
      location: {
        pathname: 'users/1/dashboard',
      },
      userRecipes: {
        userAddedRecipesCount: 0,
        userFavoritesCount: 0,
      },
      editUserProfile: jest.fn(),
      currentProfileUserId: 1,
      authenticatedUserId: 1,
    };
    const wrapper = shallow(<SideNav {...props} />);
    spy = jest.spyOn(wrapper.instance(), 'handleTabChange');

    expect(wrapper).toBeDefined();
    wrapper.find('a[name="user-recipes"]').simulate('click', {
      preventDefault: jest.fn(),
      target: {
        name: 'user-recipes'
      }
    });
    expect(wrapper).toMatchSnapshot();
  });
});
