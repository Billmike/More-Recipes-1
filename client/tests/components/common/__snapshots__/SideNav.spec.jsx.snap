// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav component should render component for authenticated user 1`] = `
<div>
  <ul
    className="side-nav fixed"
    id="slide_out"
  >
    <li>
      <Link
        replace={false}
        to="/"
      >
        Home
      </Link>
    </li>
    <li>
      <ul
        className="collapsible"
        data-collapsible="accordion"
      >
        <li>
          <div
            className="collapsible-header black-text"
          >
            Category
          </div>
          <div
            className="collapsible-body"
          >
            <ul>
              <li>
                <a
                  className="collection-item"
                  name="breakfast"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Breakfast
                </a>
              </li>
              <li>
                <a
                  className="collection-item"
                  name="lunch"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Lunch
                </a>
              </li>
              <li>
                <a
                  className="collection-item"
                  name="dinner"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Dinner
                </a>
              </li>
              <li>
                <a
                  className="collection-item"
                  name="appetizer"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Appetizer
                </a>
              </li>
              <li>
                <a
                  className="collection-item"
                  name="main"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Main
                </a>
              </li>
              <li>
                <a
                  className="collection-item"
                  name="dessert"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Dessert
                </a>
              </li>
            </ul>
          </div>
        </li>
      </ul>
    </li>
    <li>
      <div
        className="divider"
      />
    </li>
    <div>
      <li>
        <Link
          replace={false}
          to="/users/2/dashboard"
        >
          Dashboard
        </Link>
      </li>
      <li>
        <div
          className="divider"
        />
      </li>
      <div>
        <li>
          <a
            href="#!"
            name="user-profile"
            onClick={[Function]}
          >
            Profile
          </a>
        </li>
        <li>
          <a
            href="#!"
            name="user-recipes"
            onClick={[Function]}
          >
            Recipes
            <span
              className="badge"
            >
              0
            </span>
          </a>
        </li>
        <li>
          <a
            href="#!"
            name="user-favorites"
            onClick={[Function]}
          >
            Favorites
            <span
              className="badge"
            >
              0
            </span>
          </a>
        </li>
      </div>
      <div>
        <li>
          <a>
            Notifications
            <span
              className=" badge switch"
            >
              <label
                htmlFor="notification-switch"
              >
                <input
                  defaultChecked={false}
                  id="notification-switch"
                  onClick={[Function]}
                  type="checkbox"
                />
                <span
                  className="lever"
                />
              </label>
            </span>
          </a>
        </li>
        <li>
          <div
            className="divider"
          />
        </li>
      </div>
      <li>
        <a
          className="logout-button"
          onClick={[MockFunction]}
          role="button"
          tabIndex="0"
        >
          Logout
        </a>
      </li>
    </div>
  </ul>
</div>
`;

exports[`SideNav component should render component for unauthenticated user 1`] = `
<div>
  <ul
    className="side-nav null"
    id="slide_out"
  >
    <li>
      <Link
        replace={false}
        to="/"
      >
        Home
      </Link>
    </li>
    <li>
      <ul
        className="collapsible"
        data-collapsible="accordion"
      >
        <li>
          <div
            className="collapsible-header black-text"
          >
            Category
          </div>
          <div
            className="collapsible-body"
          >
            <ul>
              <li>
                <a
                  className="collection-item"
                  name="breakfast"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Breakfast
                </a>
              </li>
              <li>
                <a
                  className="collection-item"
                  name="lunch"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Lunch
                </a>
              </li>
              <li>
                <a
                  className="collection-item"
                  name="dinner"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Dinner
                </a>
              </li>
              <li>
                <a
                  className="collection-item"
                  name="appetizer"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Appetizer
                </a>
              </li>
              <li>
                <a
                  className="collection-item"
                  name="main"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Main
                </a>
              </li>
              <li>
                <a
                  className="collection-item"
                  name="dessert"
                  onClick={[MockFunction]}
                  role="button"
                  tabIndex="0"
                >
                  Dessert
                </a>
              </li>
            </ul>
          </div>
        </li>
      </ul>
    </li>
    <li>
      <div
        className="divider"
      />
    </li>
    <div>
      <li>
        <a
          onClick={[MockFunction]}
          role="button"
          tabIndex="0"
        >
          Sign In
        </a>
      </li>
      <li>
        <a
          onClick={[MockFunction]}
          role="button"
          tabIndex="0"
        >
          Register
        </a>
      </li>
    </div>
  </ul>
</div>
`;
